src/processor/LakeOfFHIR.ts

```diff
@@ -121,4 +121,36 @@ export class LakeOfFHIR implements utils.Fishable {
       );
     }
   }
+
+  /**
+   * All definitions should have a resourceType and id. If any definition is missing an id,
+   * we add one either based on the name or with a simple counter.
+   * Logs a warning if it finds any definitions without an id
+   */
+  assignMissingIds() {
+    const createdIdPaths: string[] = [];
+    let generatedId = 0;
+    this.docs.forEach((d, index) => {
+      if (d.content.id == null) {
+        // Try to be smart and set the id to the existing name
+        d.content.id = d.content.name;
```

Comment:
I think setting the `id` to the existing name is not always a foolproof way of doing things. On many resources, such as a `Patient`, or an `Organization`, the `name` does not mean the same thing as on a `StructureDefinition`. If I give as input a `Patient` without a `name` or `id` (which I think is valid FHIR @cmoesel?) then I get something that looks like this:
![image](https://user-images.githubusercontent.com/52747882/108765588-f59cda00-7521-11eb-95ae-0576196983ec.png)
Since we try to use the `name` array. I'm wondering if it may just be better not to be smart here. What are the downsides of just making every missing `id` into a generated one? I think using `name` given its inconsistent meaning in FHIR is risky.

Reply:
Good catch, @ngfreiter.  I like the idea of setting the `id` based on the name, but there are a few gotchas we didn't think of.
* As Nick noted, `name` doesn't always mean what we want it to mean.  But... if we checked if the `resourceType` was a conformance or terminology resource first, then we could more confidently use the name. We already have a list of these resource types [here](https://github.com/FHIR/GoFSH/blob/master/src/processor/InstanceProcessor.ts#L101-L117).
* That said, we'd still need to apply a little more magic to ensure the `id` fits the requirements of `id` in FHIR -- which only allows certain characters and a certain length.  We have code for that in SUSHI [here](https://github.com/FHIR/sushi/blob/master/src/fhirtypes/common.ts#L650-L679).

I guess the upside to always generating the id using a specific pattern (like `id-${count}`) is that you don't need to worry about the stuff above -- and it's easier for an author to find the ids _we_ generated if they can just search for `/id-\d+/`. So I could live with that.

Reply:
Oh this is a really good catch, thanks Nick! I think it seems feasible to do the smarter `id` setting based on `name` since it looks like we have the functions/types/tools to do this, but now I am thinking that the benefits of a consistent pattern may outweigh that "smarter" approach. This code should really only be reached if we were given invalid FHIR to begin with, which will hopefully never happen, but already has. If we just always set `id` to our pattern, it would make it clear which ids were generated. Another potential approach could be to make the generated id more clearly a GoFSH generated value, like `GOFSH-GENERATED-ID-${count}`. So I think I'm leaning toward that, but happy to do the smarter `name`-based generation if @ngfreiter or @cmoesel prefer it.

Reply:
I like `GOFSH-GENERATED-ID-${count}`.  Just to be clear, however, `id` is not required on many things, so lack of `id` doesn't necessarily indicate invalid FHIR.  But lack of `name` of a `StructureDefinition` _is_ invalid FHIR.

Reply:
Hm I wonder if I got confused after all the `name` on `StructureDefinition` thoughts from last week. But now I'm kind of reconsidering... everything. (oops). If `id` isn't required on all definitions, should I even be adding it? The root problem that sparked this PR was that `id` was `undefined` on multiple resources of the same type, so they were getting marked as duplicate and not getting processed. Is the better thing to do just ensure the `id` isn't `null` or `undefined` before checking for a duplicate to filter out? That in itself would be pretty simple, but we would hit the [check for definitions after processing](https://github.com/FHIR/GoFSH/blob/master/src/processor/Package.ts#L77) more often, but at least those resources are output.

Reply:
Oh.  Hmmm... Yeah, I think that reminds me that the reason we tried to set the `id` to the `name` (where applicable) is because if we do that, then the output FSH can suppress the `Id:` definition since SUSHI defaults `id` to the `name` anyway.  If we replace all missing ids w/ generated `id`s, then all of those will end up w/ `Id:` metadata in the FSH definitions -- and could get ugly.

So if we don't count `null`/`undefined` as a duplicate `id`, what happens after that?  Are there any other bugs/issues that arise in the GoFSH generation or no?  If not, then I might be leaning the same direction as you are, @jafeltra.

Wow, what fun we are having!

Reply:
Such fun!

The biggest change that I see when running GoFSH on some test files when using the generated `id` vs just not counting `undefined`/`null` as a duplicate `id` are a bunch of errors logged after processing that we found a resource with a duplicate name that we cannot de-duplicate. So for example, I get various errors like this:

```
error Encountered instance with a duplicate name, undefined-of-ConceptMap, which GoFSH cannot make unique. Fix the source file to resolve this error or update the resulting FSH definition.
```
These get logged in [this function](https://github.com/FHIR/GoFSH/blob/master/src/processor/Package.ts#L77) after processing, but we made the decision to add them to the package anyway, so they still get written to the output but with a funky looking name, like `undefined-of-ConceptMap`. As this branch is now, that name would either be the `name` or the `id-${count}` that we add on.

Reply:
I think we need to add `id`s to instances that don't have them, because otherwise how do we know what we are going to call that Instance when we write the FSH? My understanding is that SUSHI always generates Instances with an `id`, so no matter what we do the round tripped result won't match the original. So I vote for just generating an `id` and making it obvious it is generated.

Reply:
`undefined-of-ConceptMap` is kind of ugly, so I agree that we should generate `id`s on instances then -- but I still feel like generating `id`s on SDs, VSes, and CSes might be overstepping where we do not need to overstep (since SUSHI already has an algorithm to generate the `id` from a `name` for those cases).

Reply:
Okay, I believe that 36a9d21 makes the changes we discussed at stand up, but to summarize:

- We only try to set `id`s on Instances now. This means we only consider definitions that are not going to result in FSH Profiles, Extensions, CodeSystems, or ValueSets. Since there are some `CodeSystem` and `ValueSet` definitions that we do not support creating FSH entities for, we have to check if those definitions will be written as instances or their special FSH entity.
- If we are dealing with a definition that will be a FSH Instance, we only try to set `id` based on `name` if it is a Conformance or Terminology resource. If there is a `name`, we use that as the `id` but we change it slightly to make sure it is a valid `id`.
- If we are dealing with any other type of FSH Instance, we don't try to be smart about setting `id` and just use a GoFSH generated id with a counter.
- Once we've set `id`s on everything that is relevant to have an `id`, we can more safely exclude definitions that do not have an `id` from the duplicate definition check. This lets us process `StructureDefinitions`, `ValueSets`, and `CodeSystems` that have a `name` but no `id`.

I did have to add a handful of test fixtures that have slight variations from each other and from other existing fixtures. If anyone is bothered by having files full of slight changes and knows a way to test these cases without making those files, let me know.